### PR TITLE
[GH-725] Map configuration

### LINKED ISSUE #725: [FEATURE] Configurator WebApp: Map Configuration
We have to migrate the key "map" from `config.json`. We have to create a new map schema. It is remarkably to say that most of the keys probably are going to be associations or maps in the database. Lets discuss about it before starting this task

### PR DESCRIPTION
**Pull request by AminArria**
## Motivation

Closes #725 

## Summary of changes

- Move `map` configuration from the config JSON file to a schema in configurator app
- A decision has been made to have the inner structures as a embedded schema to simplify the UI by showing as a JSON blob

## How to test it?

Play games and take a look at: initial positions, obstacles, and spikes. All of them should be in the same spots as before (also spikes should be working)

## Checklist
- [x] Tested the changes locally.
- [ ] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: apps/arena/lib/arena/configuration.ex [hunk 1]
```diff
@@ -138,4 +150,63 @@ defmodule Arena.Configuration do
       skill
     end
   end
+
+  ## Why do we even need this? Well it happens that some of our fields are represented
+  ## as Decimal. To prevent precision loss this struct its converted to a string of the float
+  ## which should be read back and converted to Decimal
+  ## The not so small problem we have is that our code expects floats so we still need to parse
+  ## the strings, but end up with regular floats
+  defp parse_map_confg(map_config) do
```
Comment (tvillegas98) [inline apps/arena/lib/arena/configuration.ex:159]:
```suggestion
  defp parse_map_config(map_config) do
```

File: apps/arena/lib/arena/configuration.ex [hunk 2]
```diff
@@ -17,13 +17,25 @@ defmodule Arena.Configuration do
     characters = parse_characters_config(get_characters_config(), skills)
     client_config = get_client_config()
     game_config = get_game_configuration()
+    map_config = parse_map_confg(get_map_config())
```
Comment (tvillegas98) [inline apps/arena/lib/arena/configuration.ex:20]:
```suggestion
    map_config = parse_map_config(get_map_config())
```

**Review by tvillegas98 — CHANGES_REQUESTED** [review]
I came across this error when editing the initial positions

https://github.com/user-attachments/assets/1f1d494b-76a1-4c96-8a85-4ea1f2c49e41

**Review by tvillegas98 — APPROVED** [review]

**Review by tvillegas98 — DISMISSED** [review]

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/edit.html.heex [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<.header>
+  Edit Map configuration
+  <:subtitle>Use this form to manage map_configuration records in your database.</:subtitle>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/edit.html.heex:3]:
```suggestion
  <:subtitle>Use this form to manage Map Configuration records in your database.</:subtitle>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/edit.html.heex [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+</.header>
+
+<.map_configuration_form changeset={@changeset} action={~p"/map_configurations/#{@map_configuration}"} />
+
+<.back navigate={~p"/map_configurations"}>Back to map_configurations</.back>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/edit.html.heex:8]:
```suggestion
<.back navigate={~p"/map_configurations"}>Back to Map Configurations</.back>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/new.html.heex [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<.header>
+  New Map configuration
+  <:subtitle>Use this form to manage map_configuration records in your database.</:subtitle>
+</.header>
+
+<.map_configuration_form changeset={@changeset} action={~p"/map_configurations"} />
+
+<.back navigate={~p"/map_configurations"}>Back to map_configurations</.back>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/new.html.heex:8]:
```suggestion
<.back navigate={~p"/map_configurations"}>Back to Map Configurations</.back>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/new.html.heex [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<.header>
+  New Map configuration
+  <:subtitle>Use this form to manage map_configuration records in your database.</:subtitle>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/new.html.heex:3]:
```suggestion
  <:subtitle>Use this form to manage Map Configuration records in your database.</:subtitle>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex [hunk 7]
```diff
@@ -0,0 +1,57 @@
+<.header>
+  Map configuration <%= @map_configuration.id %>
+  <:subtitle>This is a map_configuration record from your database.</:subtitle>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex:3]:
```suggestion
  <:subtitle>This is a Map Configuration record from your database.</:subtitle>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  <:actions>
+    <.link href={~p"/map_configurations/#{@map_configuration}/edit"}>
+      <.button>Edit map_configuration</.button>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex:6]:
```suggestion
      <.button>Edit Map Configuration</.button>
```

File: apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        <.button type="button" phx-click={show_modal("bushes-#{@map_configuration.id}")}>Show</.button>
+        <.modal id={"bushes-#{@map_configuration.id}"}>
+          <p class="whitespace-pre"><%= embed_to_string(@map_configuration.bushes) %></p>
+        </.modal>
+      <% end %>
+    </dd>
+  </div>
+</div>
+
+<.back navigate={~p"/map_configurations"}>Back to map_configurations</.back>
```
Comment (Nico-Sanchez) [inline apps/configurator/lib/configurator_web/controllers/map_configuration_html/show.html.heex:57]:
```suggestion
<.back navigate={~p"/map_configurations"}>Back to Map Configurations</.back>
```

File: apps/configurator/test/support/fixtures/configuration_fixtures.ex [hunk 10]
```diff
@@ -1,7 +1,7 @@
-defmodule Configurator.ConfigurationFixtures do
+defmodule GameBackend.ConfigurationFixtures do
   @moduledoc """
   This module defines test helpers for creating
-  entities via the `Configurator.Configuration` context.
+  entities via the `GameBackend.Configuration` context.
```
Comment (Nico-Sanchez) [inline apps/configurator/test/support/fixtures/configuration_fixtures.ex:4]: Why this change? It's inside the configurator folder, it doesn't make much sense.

File: apps/arena/lib/arena/configuration.ex [hunk 11]
```diff
@@ -138,4 +150,63 @@ defmodule Arena.Configuration do
       skill
     end
   end
+
+  ## Why do we even need this? Well it happens that some of our fields are represented
+  ## as Decimal. To prevent precision loss this struct its converted to a string of the float
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/configuration.ex:155]: Can't we just change the Decimals to Floats instead?

**Review by Nico-Sanchez — CHANGES_REQUESTED** [review]
> A decision has been made to have the inner structures as a embedded schema to simplify the UI by showing as a JSON blob

Is this meant to be permanent? I find it very hard to read and know what the positions and obstacles are representing.
We should improve this to be more user friendly, by listing the different obstacles separately, at minimum. Then add the nested views, etc.
We can talk about this and I don't mind merging this as it's the same as main, but this shouldn't be permanent, it should be fixed inmediately after.

File: apps/configurator/test/support/fixtures/configuration_fixtures.ex (hunk 10 above)
Comment (AminArria) [reply apps/configurator/test/support/fixtures/configuration_fixtures.ex]: No idea why, but fixed

File: apps/arena/lib/arena/configuration.ex (hunk 11 above)
Comment (AminArria) [reply apps/arena/lib/arena/configuration.ex:145]: Yes, but we can't have the embedded maps because there is a sort of bug/issue with Ecto and Jason in which Floats end up been parsed back as integers (4000.0 as 4000) and that would crash while running the game cause Rust expects floats

So how about we tackle it in the follow up to fix the whole embedding situation?

**Comment by AminArria** [comment]
>Is this meant to be permanent? I find it very hard to read and know what the positions and obstacles are representing.
We should improve this to be more user friendly, by listing the different obstacles separately, at minimum. Then add the nested views, etc.
We can talk about this and I don't mind merging this as it's the same as main, but this shouldn't be permanent, it should be fixed inmediately after.

Essentially this was to have something quick, ideally we want the nesting, but the 5 level deep nesting for obstacles config means we need to think a bit on how to show this exactly, so it felt like a thing to improve on a follow up rather than get perfect now

File: apps/arena/lib/arena/configuration.ex (hunk 11 above)
Comment (Nico-Sanchez) [reply apps/arena/lib/arena/configuration.ex:145]: Yeah let's do that

**Comment by Nico-Sanchez** [comment]
> > Is this meant to be permanent? I find it very hard to read and know what the positions and obstacles are representing.
> > We should improve this to be more user friendly, by listing the different obstacles separately, at minimum. Then add the nested views, etc.
> > We can talk about this and I don't mind merging this as it's the same as main, but this shouldn't be permanent, it should be fixed inmediately after.
> 
> Essentially this was to have something quick, ideally we want the nesting, but the 5 level deep nesting for obstacles config means we need to think a bit on how to show this exactly, so it felt like a thing to improve on a follow up rather than get perfect now

Ok. Please create an issue and let's tackle that after merging this.

**Review by Nico-Sanchez — APPROVED** [review]

**Review by tvillegas98 — APPROVED** [review]
Great work!
